pkg/cloudprovider/providers/openstack/openstack_volumes.go

```diff
@@ -23,20 +23,181 @@ import (
 	"path"
 	"strings"
 
-	"k8s.io/kubernetes/pkg/volume"
+	k8s_volume "k8s.io/kubernetes/pkg/volume"
 
 	"github.com/gophercloud/gophercloud"
 	"github.com/gophercloud/gophercloud/openstack"
-	"github.com/gophercloud/gophercloud/openstack/blockstorage/v1/volumes"
+	volumes_v1 "github.com/gophercloud/gophercloud/openstack/blockstorage/v1/volumes"
+	volumes_v2 "github.com/gophercloud/gophercloud/openstack/blockstorage/v2/volumes"
 	"github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/volumeattach"
 	"github.com/gophercloud/gophercloud/pagination"
 
 	"github.com/golang/glog"
 )
 
+type IVolumes interface {
+	Create(opts VolumeCreateOpts) (string, error)
+	getVolume(diskName string) (Volume, error)
+	Delete(volumeName string) error
+}
+
+// Volumes implementation for v1
+type VolumesV1 struct {
+	blockstorage *gophercloud.ServiceClient
+	opts         BlockStorageOpts
+}
+
+// Volumes implementation for v2
+type VolumesV2 struct {
+	blockstorage *gophercloud.ServiceClient
+	opts         BlockStorageOpts
+}
+
+type Volume struct {
+	// ID of the instance, to which this volume is attached. "" if not attached
+	AttachedServerId string
+	// Device file path
+	AttachedDevice string
+	// Unique identifier for the volume.
+	ID string
+	// Human-readable display name for the volume.
+	Name string
+	// Current status of the volume.
+	Status string
+}
+
+type VolumeCreateOpts struct {
+	Size         int
+	Availability string
+	Name         string
+	VolumeType   string
+	Metadata     map[string]string
+}
+
+func (volumes *VolumesV1) Create(opts VolumeCreateOpts) (string, error) {
+
+	create_opts := volumes_v1.CreateOpts{
+		Name:         opts.Name,
+		Size:         opts.Size,
+		VolumeType:   opts.VolumeType,
+		Availability: opts.Availability,
+		Metadata:     opts.Metadata,
+	}
+
+	vol, err := volumes_v1.Create(volumes.blockstorage, create_opts).Extract()
+	return vol.ID, err
+}
+
+func (volumes *VolumesV2) Create(opts VolumeCreateOpts) (string, error) {
```

Comment:
I tend to agree that duplicate code is not good, but in this particular case, please consider why I did that way. I implement the IVolumes interface. Now, the Create/Delete methods are similar. But when v3 arrives that might not be the case. If I'll have a structure like:
```
func (volumes *VolumesV1) Create(opts VolumeCreateOpts) (string, error) {
	return generalCreate('v1',...)	
}
func (volumes *VolumesV2) Create(opts VolumeCreateOpts) (string, error) {
	return generalCreate('v2',...)	
}
````
... etc

I may force a developer, who will implement v3 support to adapt to that style, and no one knows what the lib will be like at that time. It may be not convenient.

I think that code organized this way (for this particular case) is more manageable. 

When v3 arrives IVolumes implementations may be split into separate files, eg `openstack_volumes_v#`

Did I make you change your opinion? :D